numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -1881,6 +1881,179 @@ zfill_strided_loop(PyArrayMethod_Context *context,
+
+    npy_intp N = dimensions[0];
+
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *in3 = data[2];
+    char *out1 = data[3];
+    char *out2 = data[4];
+    char *out3 = data[5];
+
+    npy_intp in1_stride = strides[0];
+    npy_intp in2_stride = strides[1];
+    npy_intp in3_stride = strides[2];
+    npy_intp out1_stride = strides[3];
+    npy_intp out2_stride = strides[4];
+    npy_intp out3_stride = strides[5];
+
+    npy_string_allocator *allocators[6] = {};
+    NpyString_acquire_allocators(6, context->descriptors, allocators);
+    npy_string_allocator *in1allocator = allocators[0];
+    npy_string_allocator *in2allocator = allocators[1];
+    // allocators[2] is NULL
+    npy_string_allocator *out1allocator = allocators[3];
+    npy_string_allocator *out2allocator = allocators[4];
+    npy_string_allocator *out3allocator = allocators[5];
+
+    PyArray_StringDTypeObject *idescr =
+            (PyArray_StringDTypeObject *)context->descriptors[0];
+    int has_string_na = idescr->has_string_na;
+    const npy_static_string *default_string = &idescr->default_string;
+
+    while (N--) {
+        const npy_packed_static_string *i1ps = (npy_packed_static_string *)in1;
+        npy_static_string i1s = {0, NULL};
+        const npy_packed_static_string *i2ps = (npy_packed_static_string *)in2;
+        npy_static_string i2s = {0, NULL};
+        npy_packed_static_string *o1ps = (npy_packed_static_string *)out1;
+        npy_packed_static_string *o2ps = (npy_packed_static_string *)out2;
+        npy_packed_static_string *o3ps = (npy_packed_static_string *)out3;
+
+        int i1_isnull = NpyString_load(in1allocator, i1ps, &i1s);
+        int i2_isnull = NpyString_load(in2allocator, i2ps, &i2s);
+
+        if (i1_isnull == -1 || i2_isnull == -1) {
+            npy_gil_error(PyExc_MemoryError, "Failed to load string in %s",
+                          ((PyUFuncObject *)context->caller)->name);
+            goto fail;
+        }
+        else if (NPY_UNLIKELY(i1_isnull || i2_isnull)) {
+            if (!has_string_na) {
+                npy_gil_error(PyExc_ValueError,
+                              "Null values are not supported in %s",
+                              ((PyUFuncObject *)context->caller)->name);
+                goto fail;
+            }
+            else {
+                if (i1_isnull) {
+                    i1s = *default_string;
+                }
+                if (i2_isnull) {
+                    i2s = *default_string;
+                }
+            }
+        }
+
+        Buffer<ENCODING::UTF8> in1buf((char *)i1s.buf, i1s.size);
+        Buffer<ENCODING::UTF8> in2buf((char *)i2s.buf, i2s.size);
+
+        // out1 and out3 can be no longer than in1, so conservatively
+        // overallocate buffers big enough to store in1
+        // out2 must be no bigger than in2
+        char *out1_mem = (char *)PyMem_RawCalloc(i1s.size, 1);
```

Comment:
We know all the sizes so in principle one can allocate the right size directly.

Though I guess the fact that one needs to allocate a buffer at all exposes a bit of a short-coming of the API: really, one would just want to allocate the output strings with the known sizes and then pass on the corresponding buffers to `string_partition`...